Shader processing changes

benaadams opened the pull request:
- Change Shaders and Programs to proper types
- Async and multi-part compile and link
- Cache and reuse vertex and fragment shader parts
- Fix and simplify disposal and cache removal
- Modularize WebGLProgram
- Retain error/warning messages, so they are available for api use
- Expose program state (compiling/linking/errored/complete etc)
- Prevent continuous error printing for failed programs (one message)

benaadams committed:
Shader processing changes

* Change Shaders and Programs to proper types
* Async and multi-part compile and link
* Cache and reuse vertex and fragment shader parts
* Fix and simplify disposal and cache removal
* Modularize WebGLProgram
* Retain error/warning messages, so they are available for api use
* Expose program state (compiling/linking/errored/complete etc)
* Prevent continuous error printing for failed programs (one message)

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -985,6 +963,12 @@ THREE.WebGLRenderer = function ( parameters ) {
 
 		var program = setProgram( camera, lights, fog, material, object );
 
+		if ( !program ) {
```
`if ( program === undefined ) return;`?

benaadams replied:
Yes, would work.

Question: Is undefined the preferred state (e.g. vs null)?

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2055,64 +2045,74 @@ THREE.WebGLRenderer = function ( parameters ) {
-		material.program = program;
+		program = material.program;
+
+		if ( program.state !== THREE.WebGLProgram.LinkedState ) {
```
This is nice. I once tried to fall back to a plain red shader for these cases. Maybe we can do that too?

benaadams replied:
Would need to be ```THREE.WebGLProgram.CompileErrorState || THREE.WebGLProgram.LinkErrorState``` as it could also be in ```THREE.WebGLProgram.LinkingState``` or ```THREE.WebGLProgram.CompilingState``` but could probably put something like that in.

Also would need to check size of position attribute to see if its 3 or 4 for the error shader.

mrdoob replied:
> Would need to be ```THREE.WebGLProgram.CompileErrorState || THREE.WebGLProgram.LinkErrorState``` as it could also be in ```THREE.WebGLProgram.LinkingState``` or ```THREE.WebGLProgram.CompilingState``` but could probably put something like that in.

Ah, true! Sounds good.

mrdoob commented:
Thanks a lot for refactoring all this!

However, I was planning on turning the `WebGLProgram` into `WebGLPrograms` just like `WebGLExtensions` (with internal cache, `.init()`, `.get()`, ...). That way you don't need `prototype` and all the `this._` stuff. Same thing with `WebGLShader` to `WebGLShaders`.

While you have all this fresh in your mind, do you mind giving this other approach a go? 😇

mrdoob commented:
`WebGLPrograms` would probably act as manager. I would still return typed `WebGLProgram`.

benaadams commented:
Pull it all out of WebGLRenderer? Am sure I can rummage something up :)

Just finishing up another PR

mrdoob commented:
Yep. Ideally everything should be out of `WebGLRenderer` 😊

benaadams committed:
Refactor shaders

Refactor webgl shaders, programs, caching
Update memory tests for async and include step time costs

benaadams commented:
@mrdoob something more like this? Still needs a bit of testing

benaadams commented:
has some bugs, working them out

benaadams commented:
Seems ok, @bhouston does this still fix #5789 for you?

benaadams committed:
Merge branch 'dev' of https://github.com/mrdoob/three.js into shaders-builds

Resolve merge conflicts

benaadams commented:
Resolved merge conflicts

bhouston commented:
Where is the async aspect? This isgoing to need extensive testing as it is a big change

benaadams commented:
> Where is the async aspect?

Alas its not async with callbacks as they don't exist; but it slightly defers checking compile and link results allowing them more time to complete before forcing a js pause to wait for the results (as per @toji's [Asynchronous" WebGL shader compiling](https://twitter.com/Tojiro/status/584820688772431872) )

In both current and new whole compiled program with same parameters *exists*, that is used, however when it doesn't the logic is changed. The new works more like textures do, ish.

Whole compiled program with same parameters *does not exist*

##### Current Logic

1. Create and Compile vertex shader on GPU
2. Check results (wait for completion)
 * Output Errors if occurred
3. Create and Compile fragment shader on GPU
4. Check results (wait for completion)
 * Output Errors if occurred
5. Create and Link program on GPU 
6. Check results (wait for completion)
 * Output Errors if occurred
7. Query uniform and attribute locations.
8. Return program
 * Output continuous errors if error occurred
9. Render object

##### New Logic

1. Check if matching vertex shader exists in cache
 * Use if does
 * If not create on GPU and submit for compile, mark as compiling, add to cache
2. Check if matching fragment shader exists
 * Use if does
 * If not create on GPU and submit for compile, mark as compiling, add to cache
3. Check if shaders are pre-compiled (from cache), if so move to (7. Link step), if from cache but not compiled move to (5. Check results), else return WebGLProgram reference then:
4. Skip object render -> new frame or next object using same shaders
5. Check vertex shader results (wait for completion)
 * Output Errors if occurred, mark shader as errored, mark program as errored
 * Else mark as compiled
6. Check fragment shader results (wait for completion)
 * Output Errors if occurred, mark shader as errored, mark program as errored
 * Else mark as compiled
7. (Link step) Create program on GPU, submit for linking, mark program as Linking
8. Skip object render -> new frame or next object using same program
9. Check Link results (wait for completion)
 * Output Errors if occurred, mark program as errored
 * Else mark as linked
10. Query uniform and attribute locations.
11. Render object

With the new logic it can take 3 attempts at using the program for it to be ready (either 3 objects using same shaders or 3rd frame); but the old logic will instead block other js from running while it waits.

The new logic also caches the shader parts (vertex+fragment); so if you have 15 different programs but the vertex shader is the same in all of them it only is compiled once and that step is skipped for the other 14 programs; whereas previously it would be compiled 15 times.

Error handling is different and both errored shaders and programs are cached and marked as errorred; so if something else uses them it doesn't try to do anything with them as they are immediately identified as errored; with the error being previously reported. 

Probably not very clear, but hope it makes some sense?

benaadams commented:
@bhouston specificly in code it is here: https://github.com/mrdoob/three.js/pull/6370/files#diff-5e43a0b5002eb2c419def3baf67d4e67R31

benaadams commented:
A test which shows this is the webgl_test_memory2 (http://prealpha.ageofascent.com/examples/webgl_test_memory2.html) which now outputs more information to the console:

1: compile submit  - meshes: 100 vertex: 1  fragment: 100  programs: 0  time: 7.8 ms
2: wait compile complete + link submit  - meshes: 100 vertex: 1  fragment: 100  programs: 100  time: 41.8ms
3: wait link complete + draw  - meshes: 100 vertex:  1  fragment:  100  programs:  100  time: 906.6 ms
4: just draw  - meshes: 100 vertex:  1  fragment:  100  programs:  100  time: 0.9 ms
5: programs disposed - meshes: 100 vertex:  0  fragment:  0  programs:  0  time: 0.4 ms

The times don't include extra elapsed between rAF so things take longer than the times above; just the js engine is able to continue executing rather than being paused.

gero3 commented:
If I read this right, you need to render 3 times to get a result. I don't believe this is recommended. 
None the less, I like the idea of async processing of the shaders. To accomplish this, I think we should start with compiling all shaders that needs compiling and when this is done, then check for errors and then go further by drawing the nessecary objects.

benaadams commented:
> If I read this right, you need to render 3 times to get a result.

Only when introducing a shader for the first time; much like the textures don't render until they are loaded but the frames still draw (unless overridden)

>  I think we should start with compiling all shaders that needs compiling and when this is done, then check for errors and then go further by drawing the nessecary objects.

Well... Also load all buffers and textures at start of frame in order, and then move through to draw, so something like:

1. Compile
2. Load buffers
3. Link
4. Load textures
5. Draw

Or at end of frame:

1. Draw
2. Compile
3. Load buffers
4. Link
5. Load textures

-- data ready for next frame --

With the step order being configurable, so also if happy to do 30fps

-- odd frame --
1. Draw

-- even frame --
1. Compile
2. Load buffers
3. Link
4. Load textures

-- odd frame --
1. Draw

-- even frame --
1. Compile... etc

But I was going to follow up with that after this PR :)

benaadams commented:
@gero3 was doing something similar in PR #5954 though it was a bit ugly

bhouston commented:
@benaadams You know there is a different way of coding renders that actually I find simpler to write and maintain, but they are more complex to use.  Basically you can do low-level batching renderer.  I wrote a functional prototype in JavaScript based on some old C# code I had that did this:

The renderer:

https://gist.github.com/bhouston/c943d02533ab04e53dc7

A test that shows how you use it:

https://gist.github.com/bhouston/6426cbd6106010f34fba

For maximum performance and control I prefer the above approach, but it conflicts with the ease of use of ThreeJS's current structure.  My suggestion would be to move towards this type of design with ThreeJS's current structure as a nice layer on top that makes it more usable.  But I think for advanced users, such as yourself and myself, I would prefer to use the low level batch renderer directly.

PS. One thing I wanted to add to my prototype, but didn't was buffer reference counting so that I could share buffers between different batch items.

bhouston commented:
@benaadams The basic idea is that everything is buffers and programs and uniform sets.  And you actually work at that level.  It leads to massive efficiencies, both in terms of the size of the framework, and also one's code and it is simpler to think about too.

My prototype is very very rough though.  And it is 16 months old at this point.

bhouston commented:
@benaadams The idea is that we have a low level renderer that doesn't really understand the uniforms at all, that is left up to the user.  It doesn't know about lights, or materials or anything.  It just does updates and batches as it is commanded to.

And I think a lot of the refactoring @mrdoob is doing is going somewhat in this direction.

benaadams commented:
@bhouston interesting! Maybe should wrap this up as a bigger restructure with better batched pipelining? Though my worry on bigger changes is merge conflicts.

@gero3 to follow up on your earlier point; its it preferable not to render anything until all shaders are ready over rendering things as the shaders become available; with possibility a 3 frame wait? I could see it being a problem for gpgpu or items that only render a single frame rather than animating.

bhouston commented:
> its it preferable not to render anything until all shaders are ready over rendering things as the shaders become available; with possibility a 3 frame wait?

I'd prefer that the engine not make decisions like that.

benaadams commented:
> I'd prefer that the engine not make decisions like that.

Ok, will put the pipelining/batching in similar to what @bhouston was doing with batch renderer, api should remain the same.

Hold on to your hats... (other work depending, though I have learnt the pain of trying to fork back after deviating for a year so hopefully won't be as long this time!)

benaadams commented:
There's an awful amount of code in WebGLRenderer :(

benaadams committed:
Phases

Only update params (lights/shadows) once per scene
Compile phase
Buffer update phase
Link phase
Texture update phase
Link Complete phase
Draw phase

benaadams commented:
Still some to do, but does do some batch phases now.

1. Compile (If needed)
2. Buffer Upload (if needed)
3. Link (if needed)
4. Draw

For the scene as a whole rather than per object

mrdoob commented:
There are too many changes in this PR. Would it be possible to split it in a bunch of PR?
I've noticed a bunch of bugs got introduced lately because of this multi-feature PR approach.

benaadams commented:
Yep

RileyLee commented:
Is there a way I can avoid recompiling a shader program but instead store it somewhere? So it would be faster the next time when I want to switch back from a different shader.
